private/bufpkg/bufpluginexec/wasm_handler.go

```diff
@@ -0,0 +1,162 @@
+	))
+	defer span.End()
+	if enabled, err := IsAlphaWasmEnabled(container); enabled == false {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return err
+	}
+	requestData, err := protoencoding.NewWireMarshaler().Marshal(request)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return err
+	}
+	pluginBytes, err := os.ReadFile(h.pluginPath)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return err
+	}
+	compiledPlugin, err := h.wasmPluginExecutor.CompilePlugin(ctx, pluginBytes)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return err
+	}
+	responseBuffer := bytes.NewBuffer(nil)
+	if err := h.wasmPluginExecutor.Run(
+		ctx,
+		compiledPlugin,
+		// command.RunWithEnv(app.EnvironMap(container)), // TODO, not exposed right now
+		bytes.NewReader(requestData),
+		responseBuffer,
+	); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		if pluginErr := new(bufwasm.PluginExecutionError); errors.As(err, &pluginErr) {
+			_, _ = container.Stderr().Write([]byte(pluginErr.Stderr))
+		}
+		return err
+	}
+	response := &pluginpb.CodeGeneratorResponse{}
+	if err := protoencoding.NewWireUnmarshaler(nil).Unmarshal(responseBuffer.Bytes(), response); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return err
+	}
+	response, err = normalizeCodeGeneratorResponse(response)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		return err
+	}
+	if response.GetSupportedFeatures()&uint64(pluginpb.CodeGeneratorResponse_FEATURE_PROTO3_OPTIONAL) != 0 {
+		responseWriter.SetFeatureProto3Optional()
+	}
+	for _, file := range response.File {
+		if err := responseWriter.AddFile(file); err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+			return err
+		}
+	}
+	// plugin.proto specifies that only non-empty errors are considered errors.
+	// This is also consistent with protoc's behavior.
+	// Ref: https://github.com/protocolbuffers/protobuf/blob/069f989b483e63005f87ab309de130677718bbec/src/google/protobuf/compiler/plugin.proto#L100-L108.
+	if response.GetError() != "" {
+		responseWriter.AddError(response.GetError())
+	}
+	return nil
+}
+
+func validateWASMFilePath(path string) (string, error) {
+	path, err := filepath.Abs(path)
+	if err != nil {
+		return path, err
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return path, err
+	}
+	if info.IsDir() || info.Size() == 0 || !strings.HasSuffix(path, ".wasm") {
+		return path, fmt.Errorf("invalid WASM file: %s", path)
+	}
+	return path, nil
+}
+
+// IsAlphaWasmEnabled returns if BUF_ALPHA_ENABLE_WASM is set to true.
+func IsAlphaWasmEnabled(container app.EnvContainer) (bool, error) {
+	enabled, _ := container.GetEnvBoolValue(AlphaEnableWasmEnvKey)
+	if !enabled {
+		return false, errors.New("alpha wasm support is disabled")
```

Comment:
One of these values doesn't matter, I would either hoist the error up to the caller, or drop the bool and rename this to `RequireAlphaWasm(app.EnvContainer) error`.

Reply:
Clarified and updated offline.